File: config/core/schemas/soil.py
"""
KisanAI OS
Soil Schemas
Version: 5.3.0
"""

from typing import Optional

from pydantic import BaseModel, Field, field_validator


def _clean_required(value):
    """Strip + collapse whitespace; reject blank input."""
    if value is None:
        raise ValueError("value is required")

    text = " ".join(str(value).strip().split())

    if not text:
        raise ValueError("value cannot be blank")

    return text


class SoilCreate(BaseModel):
    soil_id: Optional[int] = None
    farmer_id: Optional[int] = None
    soil_type: str
    ph: float = Field(ge=0.0, le=14.0)
    moisture: float = Field(ge=0.0, le=100.0)
    nitrogen: int = Field(ge=0)
    phosphorus: int = Field(ge=0)
    potassium: int = Field(ge=0)

    @field_validator("soil_type", mode="before")
    @classmethod
    def _normalize_soil_type(cls, value):
        return _clean_required(value)


class SoilUpdate(BaseModel):
    farmer_id: Optional[int] = None
    soil_type: str
    ph: float = Field(ge=0.0, le=14.0)
    moisture: float = Field(ge=0.0, le=100.0)
    nitrogen: int = Field(ge=0)
    phosphorus: int = Field(ge=0)
    potassium: int = Field(ge=0)

    @field_validator("soil_type", mode="before")
    @classmethod
    def _normalize_soil_type(cls, value):
        return _clean_required(value)


class SoilOut(BaseModel):
    soil_id: int
    farmer_id: Optional[int] = None
    soil_type: str
    ph: float
    moisture: float
    nitrogen: int
    phosphorus: int
    potassium: int
    created_at: str

File: mobile/lib/models/soil.dart
/// Soil record model (`SoilOut`).
class Soil {
  const Soil({
    required this.soilId,
    required this.farmerId,
    required this.soilType,
    required this.ph,
    required this.moisture,
    required this.nitrogen,
    required this.phosphorus,
    required this.potassium,
    required this.createdAt,
  });

  final int soilId;
  final int? farmerId;
  final String soilType;
  final double ph;
  final double moisture;
  final int nitrogen;
  final int phosphorus;
  final int potassium;
  final String createdAt;

  factory Soil.fromJson(Object? json) {
    final map = json as Map<String, dynamic>;
    return Soil(
      soilId: (map['soil_id'] as num).toInt(),
      farmerId: (map['farmer_id'] as num?)?.toInt(),
      soilType: map['soil_type'] as String? ?? '',
      ph: (map['ph'] as num?)?.toDouble() ?? 0,
      moisture: (map['moisture'] as num?)?.toDouble() ?? 0,
      nitrogen: (map['nitrogen'] as num?)?.toInt() ?? 0,
      phosphorus: (map['phosphorus'] as num?)?.toInt() ?? 0,
      potassium: (map['potassium'] as num?)?.toInt() ?? 0,
      createdAt: map['created_at'] as String? ?? '',
    );
  }
}
